src/mbgl/renderer/renderer_impl.cpp

```diff
@@ -369,6 +369,47 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
         parameters.frameHistory.upload(parameters.context, 0);
     }
 
+    // - PREPARE + CLIP ------------------------------------------------------------------------------
+    {
+        MBGL_DEBUG_GROUP(parameters.context, "clip");
+
+        // Update all clipping IDs.
+        for (const auto& entry : renderSources) {
+            if (entry.second->isEnabled()) {
+                entry.second->startRender(parameters);
+            }
+        }
+    }
+
+    // - 3D PASS -------------------------------------------------------------------------------------
+    // Renders any 3D layers bottom-to-top to unique FBOs with texture attachments, but share the same
+    // depth rbo between them.
+    {
+        MBGL_DEBUG_GROUP(parameters.context, "3d");
+        parameters.backend.bind();
+        parameters.pass = RenderPass::Pass3D;
+
+        const auto size = parameters.context.viewport.getCurrentValue().size;
+
+        if (!parameters.staticData.depthRenderbuffer ||
+            parameters.staticData.depthRenderbuffer->size != size) {
+            parameters.staticData.depthRenderbuffer =
+                parameters.context.createRenderbuffer<gl::RenderbufferType::DepthComponent>(size);
```

Comment:
Before creating any objects here, we should first check if there are even extrusion layers in the stylesheet. Otherwise, all of those resources + backend binding won't be necessary.